File: clients/algoliasearch-client-go/algolia/transport/request_id.go
package transport

import (
	"crypto/rand"
	"strings"
	"time"
)

// RequestIDHeader is the name of the header carrying the Request-ID minted
// when Configuration.RequestIDEnabled is set.
const RequestIDHeader = "Request-ID"

const requestIDAlphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"

const requestIDLength = 11

// NewRequestID returns a fresh 11-character base62 identifier suitable for
// the Request-ID header. The modulo bias of the byte mapping is acceptable:
// the ID is a tracing breadcrumb, not a secret.
func NewRequestID() string {
	b := make([]byte, requestIDLength)

	_, err := rand.Read(b)
	if err != nil {
		// A request must not fail over its tracing metadata: when the entropy
		// source is unavailable, degrade to a time-seeded sequence.
		n := time.Now().UnixNano()
		for i := range b {
			n = n*6364136223846793005 + 1442695040888963407
			b[i] = byte((n >> 33) & 0xff)
		}
	}

	for i := range b {
		b[i] = requestIDAlphabet[int(b[i])%len(requestIDAlphabet)]
	}

	return string(b)
}

// HasRequestID reports whether any of the given header maps carries a
// Request-ID entry, whatever its casing. Unlike http.Header, plain header
// maps keep the caller's literal casing, so the lookup must not assume a
// canonical form.
func HasRequestID(headers ...map[string]string) bool {
	for _, m := range headers {
		for k := range m {
			if strings.EqualFold(k, RequestIDHeader) {
				return true
			}
		}
	}

	return false
}

File: clients/algoliasearch-client-swift/Sources/Core/Networking/RequestID.swift
//
//  RequestID.swift
//
//
//  Created by Algolia on 10/08/2026.
//

import Foundation

/// Mints the Request-ID tracing header sent by the clients that support it.
public enum RequestID {
    /// The name of the header carrying the Request-ID.
    public static let httpHeaderField = "Request-ID"

    private static let alphabet: [Character] = Array(
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
    )

    /// Returns a fresh 11-character base62 identifier suitable for the Request-ID header.
    public static func generate() -> String {
        String((0 ..< 11).compactMap { _ in self.alphabet.randomElement() })
    }

    /// Whether the given headers already carry a Request-ID entry, whatever its casing.
    /// Header dictionaries keep the caller's literal casing, so the lookup must not assume
    /// a canonical form.
    public static func hasRequestID(in headers: [String: String]?) -> Bool {
        guard let headers else {
            return false
        }

        return headers.keys.contains { $0.caseInsensitiveCompare(self.httpHeaderField) == .orderedSame }
    }

    /// Derives the request options carrying the Request-ID shared by every request of one
    /// helper invocation. Returns the options untouched when the configuration does not
    /// support Request-ID or the caller already supplied one through the options or the
    /// default headers, which also makes nested helpers reuse the ID minted by their caller.
    public static func withRequestID(
        _ requestOptions: RequestOptions?,
        configuration: BaseConfiguration
    ) -> RequestOptions? {
        guard configuration.requestIDEnabled,
              !self.hasRequestID(in: requestOptions?.headers),
              !self.hasRequestID(in: configuration.defaultHeaders)
        else {
            return requestOptions
        }

        return RequestOptions(headers: [self.httpHeaderField: self.generate()]) + requestOptions
    }
}
